File: packages/lynx-http/android/com/sigx/http/HttpPublisher.kt
package com.sigx.http

import android.util.Log
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.tasm.LynxView
import java.util.UUID

/**
 * Per-[LynxView] publisher that pumps [HttpEventBus] payloads into JS via
 * `LynxView.sendGlobalEvent("__sigxHttpEvent", [...])`.
 *
 * One instance per LynxView; instantiated by the generated
 * `GeneratedLifecyclePublishers.attachAll(lynxView)` and retained for the
 * LynxView's lifetime — same wiring as `WebSocketPublisher`.
 */
class HttpPublisher(private val lynxView: LynxView) {

    private var token: UUID? = null

    fun attach() {
        token = HttpEventBus.addListener { json ->
            try {
                // Pass a single JSON string as the only param — the JS shim
                // parses the first arg as `NativeHttpEvent`. A string survives
                // Lynx 0.5.0's bridge marshalling intact where a structured
                // map carrying a nested `headers` map does not (#342).
                val params = JavaOnlyArray()
                params.pushString(json)
                lynxView.sendGlobalEvent(EVENT_NAME, params)
            } catch (e: Throwable) {
                Log.w(TAG, "publish failed: ${e.message}")
            }
        }
    }

    fun detach() {
        token?.let { HttpEventBus.removeListener(it) }
        token = null
    }

    private companion object {
        const val TAG = "SigxHttpPublisher"
        const val EVENT_NAME = "__sigxHttpEvent"
    }
}

File: packages/lynx-http/ios/HttpPublisher.swift
import Foundation
import Lynx

/// Per-`LynxView` publisher that pumps `HttpEventBus` payloads into JS via
/// `LynxView.sendGlobalEvent("__sigxHttpEvent", [...])`.
///
/// One instance per LynxView; instantiated by the generated
/// `GeneratedLifecyclePublishers.attachAll(to:)` and retained for the
/// LynxView's lifetime — same wiring as `WebSocketPublisher`.
final class HttpPublisher {

    private weak var lynxView: LynxView?
    private var token: UUID?

    init(lynxView: LynxView) {
        self.lynxView = lynxView
        self.token = HttpEventBus.shared.addListener { [weak self] json in
            guard let view = self?.lynxView else { return }
            // sendGlobalEvent expects an array of params; pass a single JSON
            // string as the only param — the JS shim parses the first arg as
            // `NativeHttpEvent`. A string (rather than a structured map)
            // survives Lynx 0.5.0's bridge marshalling intact (see #342 /
            // `HttpEventBus.emit`).
            view.sendGlobalEvent("__sigxHttpEvent", withParams: [json])
        }
    }

    deinit {
        if let token = token {
            HttpEventBus.shared.removeListener(token)
        }
    }
}
